rfcs/2021-10-15-9572-accept-datadog-traces.md

```diff
@@ -0,0 +1,221 @@
+[short description of APM events]: https://github.com/DataDog/datadog-agent/blob/e081bed/pkg/trace/event/doc.go
+
+## Cross cutting concerns
+
+[On-going work][schema-rfc] to support event schema would allow to express some constrains on an event structure. In
+this case this would allow to formalize a trace schema while keeping the underlying data as standard Vector event. The
+trace sink would then expect event following this schema.
+
+[schema-rfc]: https://github.com/vectordotdev/vector/pull/9388
+
+## Scope
+
+### In scope
+
+* Ingest traces from the trace-agent in the `datadog_agent` source
+* Send traces to the Datadog trace endpoint through a new `datadog_trace` sink
+* Basic operation on traces: filtering, routing
+* Pave the way for OpenTelemetry traces
+
+### Out of scope
+
+* Profiling data, but as the trace-agent only proxies profiling data, the same behaviour can be implemented rather
+  quickly in Vector.
+* Debugger logs can be already diverted to Vector (untested but it should work as Vector supports datadog logs and there
+  is an config option to explicitly configure the debugger log destination)
+* Metrics emitted by the trace-agent (they could theoretically be received by Vector by a statsd source, but the host
+  used by the trace-agent is derived from the local config to programmatically discover the main agent, thus there is no
+  existing knob to force the trace agent to send metrics to a custom dogstatsd host)
+* Span extraction, filtering
+* Other sources & sinks for traces than `datadog_agent` source & `datadog_trace` sink
+
+## Pain
+
+Vector does not support any traces (full json representation may be ingested as log event) at the moment and it is a
+key part of observability. Therefore, users cannot use Vector for the business-level user cases on trace data, like
+cost control and reduction, redacting PII, routing, and more.
+
+## Proposal
+
+### User Experience
+
+* User will be able to ingest traces from the trace agent
+  * Vector config would then consist of: `datadog_agent` source -> some filtering/enrichment transform ->
+    `datadog_trace` sink
+  * Datadog trace agent can be configured to send traces to any arbitrary endpoint using `apm_config.apm_dd_url` [config
+    key]
+* This change is a pure addition to Vector, there will be no impact on existing feature
+
+[config key]: https://github.com/DataDog/datadog-agent/blob/34a5589/pkg/config/apm.go#L61-L87
+
+### Implementation
+
+To keep vector-core as generic as possible, the first implementation will decode datadog traces as `LogEvent`, the
+resulting event will be deeper than usual but this should not be a problem. To keep track that those `LogEvent`s are
+traces, we could add a simple `Option<bool>` to the `EventMetadata` struct. However this may not be immediately
+required. While being naive this does not predates future vector evolution towards event traits by not adding a new
+concrete type into the `Event` enum.
```

Comment:
I was chatting with @lukesteensen this morning about this and I think there may be some confusion here about the modeling. We were thinking that we _would_ add a new enum type, `Trace`, to `Event` and just have it wrap `LogEvent`. That is:

```rust
enum Event {
  // ....
  Trace(LogEvent)
}
```

This would allow us to treat traces as a first-class type in Vector, but reuse most of the `LogEvent` implementation, for now at least.